File: Benchmarks_FSharp/Core/Components/TransformComponent.fs
namespace Benchmarks.Core.Components

open Benchmarks.Core.Types

[<Struct>]
type TransformComponent =
    val mutable Size: Vector3
    val mutable Scale: float32
    val mutable Rotation: float32
    val mutable Position: Vector3
    val mutable SpriteEffects: SpriteEffects
    val mutable Color: Color

    new(position: Vector3, size: Vector3, scale: float32, spriteEffects: SpriteEffects, color: Color) =
        { Position = position; Size = size; Scale = scale; Rotation = 0f; SpriteEffects = spriteEffects; Color = color }

    new(position: Vector3, size: Vector3) =
        { Position = position; Size = size; Scale = 1f; Rotation = 0f; SpriteEffects = SpriteEffects.None; Color = Color.White }

    new(position: Vector3, size: Vector3, scale: float32) =
        { Position = position; Size = size; Scale = scale; Rotation = 0f; SpriteEffects = SpriteEffects.None; Color = Color.White }

    member this.SetPosition(x: float32, y: float32, z: float32) =
        this.Position <- Vector3(x, y, z)

File: Benchmarks/Core/Components/TransformComponent.cs
using Benchmarks.Core.Types;

namespace Benchmarks.Core.Components
{
    public struct TransformComponent(Vector3 position, Vector3 size, float scale, SpriteEffects spriteEffects, Color color)
    {
        public void SetPosition(float x, float y, float z) => Position = new Vector3(x, y, z);

        public Vector3 Size { get; set; } = size;
        public float Scale { get; set; } = scale;
        public float Rotation { get; set; } = 0f;
        public Vector3 Position { get; set; } = position;
        public SpriteEffects SpriteEffects { get; set; } = spriteEffects;
        public Color Color { get; set; } = color;

        public TransformComponent(Vector3 position, Vector3 size) : this(position, size, 1f, SpriteEffects.None, Color.White) { }
        public TransformComponent(Vector3 position, Vector3 size, float scale) : this(position, size, scale, SpriteEffects.None, Color.White) { }
    }
}
